torchaudio/prototype/functional/functional.py

```diff
@@ -76,3 +76,61 @@ def convolve(x: torch.Tensor, y: torch.Tensor) -> torch.Tensor:
     )
     output_shape = x.shape[:-1] + (-1,)
     return output.reshape(output_shape)
+
+
+def add_noise(waveform: torch.Tensor, noise: torch.Tensor, lengths: torch.Tensor, snr: torch.Tensor) -> torch.Tensor:
+    r"""Scales and adds noise from multiple sources to waveform per signal-to-noise ratio.
+
+    Specifically, for each waveform vector :math:`x \in \mathbb{R}^L` and noise vectors
+    :math:`n_1, \ldots, n_N \in \mathbb{R}^L` corresponding to :math:`N` sources, the
+    function computes output :math:`y` as
+
+    .. math::
+        y = x + \sum_{i = 1}^N a_i n_i
+
+    , where
+
+    .. math::
+        a_i = \sqrt{ \frac{ ||x||_{2}^{2} }{ ||n_i||_{2}^{2} } \cdot 10^{-\frac{\text{SNR}_i}{10}} }
+
+    , with :math:`\text{SNR}_i` being the desired signal-to-noise ratio between :math:`x` and :math:`n_i`, in dB.
+
+    Note that this function broadcasts singleton leading dimensions in its inputs in a manner that is
+    consistent with the above formulae and PyTorch's broadcasting semantics.
+
+    Args:
+        waveform (torch.Tensor): Input waveform, with shape `(*, L)`.
+        noise (torch.Tensor): Noise, with shape `(*, N, L)` (leading dimensions must match those of ``waveform``).
+        lengths (torch.Tensor): Valid lengths of signals in `waveform` and `noise`, with shape `(*,)`.
+        snr (torch.Tensor): Signal-to-noise ratios in dB, with shape `(*, N)`.
+
+    Returns:
+        torch.Tensor: Result of scaling and adding ``noise`` to ``waveform``, with shape `(*, L)`
+        (same shape as ``waveform``).
+    """
+
+    input_leading_dims = (len(waveform.shape[:-1]), len(noise.shape[:-2]), len(lengths.shape), len(snr.shape[:-1]))
+    if any([a != input_leading_dims[0] for a in input_leading_dims[1:]]):
+        raise ValueError(f"Input leading dimensions don't match (got {input_leading_dims}).")
+
+    if waveform.shape[-1] != noise.shape[-1]:
+        raise ValueError(f"Length dimensions of waveform and noise don't match (got {waveform[-1]} and {noise[-1]}).")
+
+    if noise.shape[-2] != snr.shape[-1]:
+        raise ValueError(f"Noise source dimensions of noise and snr don't match (got {noise[-2]} and {snr[-1]}).")
+
+    # compute scale
+    mask = torch.arange(0, waveform.size(-1)).expand(waveform.shape) < lengths.unsqueeze(-1)  # (*, L) < (*, 1) = (*, L)
+    energy_signal = torch.linalg.vector_norm(waveform * mask, ord=2, dim=-1) ** 2  # (*,)
+    energy_noise = torch.linalg.vector_norm(noise * mask.unsqueeze(-2), ord=2, dim=-1) ** 2  # (*, N)
+    original_snr = energy_signal.unsqueeze(-1) / energy_noise  # (*, N)
+    snr_abs = 10 ** (snr / 10.0)  # (*, N)
+    scale = (original_snr / snr_abs).sqrt()  # (*, N)
+
+    # scale noise
+    scaled_noise = scale.unsqueeze(-1) * noise  # (*, N, 1) * (*, N, L) = (*, N, L)
+
+    # sum-reduce scaled noise
+    scaled_noise = scaled_noise.sum(-2)  # (*, L)
```

Comment:
I feel that letting client code do this will make this feature more flexible.
Not sure how common to add noises like this.

Reply:
a output of shape `(*, N, L)` that comprises individual sums of input waveform and scaled noise waveform would be difficult to reduce in client code — you'd need to do something like

`output.sum(-2) - (N - 1) * waveform`

to produce the current return value.

alternatively (or additionally), we could introduce a `scale_noise` function that simply returns noise scaled in accordance with the given waveform and snr

Reply:
computing `SNR` for each noise source then summing the noises is quite tricky, usually when we talk about `SNR`, it means the ratio of speech signal to all noises in the mixture. If users provide a `SNR` Tensor for 4 sources, for example. If the `SNR` values are [1, 2, -3, 0], after the mixing, none of the `SNR` values represent the real SNR of the mixture.
Shall we make a note in the method to let users be aware of it, or can we return the final `SNR` score for the mixture?

Reply:
> none of the SNR values represent the real SNR of the mixture.

This is a very a good point. I think at least we should have a function that returns the mixture with the specified SNR.
Other logics can be built on top of it.

Reply:
as it stands, the snr is defined per noise source, not on each waveform, so returning a "mixture with the specified snr" is not possible. and if it were instead defined on each waveform, a unique solution for the composition of noise sources to achieve the desired snr wouldn't exist

Reply:
How about making a core function take the form of `(waveform [N, L], noise [N, L], snr [N, ], length[N]) -> mixture [N, L]`, and build features with more elaborated logic using the core function?

Reply:
revised to eliminate hardcoded noise source dimension — users can account for multiple noise sources in one of the leading dimensions if they want